ci: Github Actions workflows for publish to marketplace (#4606)
forcedotcom/salesforcedx-vscode: contributing/publishing.md

## Merging the Release Branch into Main

After the change log has been approved and merged into your release branch, it's time to prepare main with the new changes for the publish. A Github Action workflow is exectued to merge the release branch. We are specifically using the rebase strategy because we want all the commits from our release branch to be applied on top of the commits in main.

### To run the merge process:

1. From the github repository navigate to the Action Tab and Select the PrePublish workflow on the left
1. Click the 'Run Workflow' button and input the branch to be merged into the input box.
1. Click the `Run Workflow` button.

The Prepublish job will verify the version of the branch to be merged is newer then what is currently in main and update main with the release branch.

## Publishing Main

The merge into main will trigger a run of the 'Test, Build, and Publish' GHA workflow that will run the tests, build vsix files, and send a slack notification that a publish workflow has been initiated.

> After all tests have passed it will trigger a required approval for continueing the job.

Reviewer: ```suggestion
After all tests have passed it will trigger a required approval for continuing the job.
```

> Before approving the publish to the marketplace download the vsix files, install them locally and verify they are working as expected.

Reviewer 2: 👍🏽

# Manual Publish

> The steps used to execute a publish to the marketplace can be found in the assoicated github actions.

Reviewer: ```suggestion
The steps used to publish to the VS Code Marketplace can be found in the associated Github Actions.
```

## Generating a Major Release

The versioning we follow is intentionally mapped with Salesforce Core. When a major version bump occurs, such as 53.0 -> 54.0, we release a major version update as well.

## Downloading the .vsix from Github Action

### Options

> - Download directly from the Github Action run. You will find artifacts that are assoicated with a run at the bottom of the summary screen

Reviewer: ```suggestion
- Download directly from the Github Action run. You will find artifacts that are associated with a run at the bottom of the summary screen
```

- Use the gh cli to download artifacts. `gh run download --dir /dir/where/you/want/the/vsix/files/ 3746978326`. The last arg is the GHA job id. This can be found in the UI or by executing `gh run list`.

**At this stage, it is possible to share the .vsix directly for manual
installation.**

To manually install vsix files you can use the `code` or `code-insiders` cli.

- `code-insiders --install-extension /path/to/the/vsix/iama.vsix`

> - or install all downloaded vsix files `find ./vsix/download/path -type f -name "*.vsix" -exec code --install-extension {} \;`

Reviewer 2: 💯

> 

Author: I outlined how you can download the vsix artifacts from GHA here and how to easily install those.  let me know if you think we need an in between step for something closer to ctv, but it seems pretty straightforward.
